### PR TITLE
Metadata File Import UI

### LINKED ISSUE #357: Project Sample Metadata - File Import UI
## Summary

As a user I want to be able to import metadata for a batch of samples from a file. csv, excel files should be supported. I want the option to ignore empty values, which would mean the metadata fields with empty values are not deleted from the sample metadata

## Acceptance Criteria

- Create a new modal which accomplishes the following:
1) User selects a metadata file to upload (csv, excel format, etc)
2) Stimulusjs controller picks up that a file was selected and does a quick parse of the first row to get the header columns
3) A dropdown is populated with the header fields from step 2
4) A checkbox is visible to the user to select for `Ignore empty values`. Selecting this box will ensure that those fields are not removed from the metadata as the default behaviour is for fields with empty values to be removed
5) A `Import metadata` button which is disabled unless a field is selected in step 3
6) Response should be displayed in a modal as there could be quite a bit of info to display such as if there are failures with the upload. This modal will need a close/ok button for the user to dismiss the modal

### PR DESCRIPTION
**Pull request by ksierks**
## What does this PR do and why?
Created a new metadata file import modal.
Fixes #357

## Screenshots or screen recordings
image

Successful import:
image

File validation errors:
image

Missing sample errors:
image

## How to set up and validate locally
1. Navigate to a project samples page that the logged in user has access to update.
2. Hit the `Import Metadata` button.
3. Select a file.
4. Select the `Sample ID Column`. (Note: Currently finds samples by name.)
5. Hit the `Import Metadata` button.
6. Verify the sample metadata was updated.
7. Try with and without ignoring empty values.

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [ ] I have evaluated the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.

### CODE REVIEW

**Review by joshsadam — CHANGES_REQUESTED** [review]
Just taking a quick look through and found a couple issues I would like you to take a look at.  It does seem to function well.

File: app/javascript/controllers/projects/samples/metadata/file_import_controller.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import { Controller } from "@hotwired/stimulus";
+import * as XLSX from "xlsx";
+
+export default class extends Controller {
+  static targets = ["fileInput", "selectInput", "submitButton"];
+
+  connect() {
+    this.fileInputTarget.addEventListener("change", (event) => {
```
[the remark this answers is not in the document]
Comment (ksierks) [reply app/javascript/controllers/projects/samples/metadata/file_import_controller.js]: My bad. That'll make the controller smaller. Please see 7ee3e6dcf665f164e6f80d2089ad7a78258fcdff.

File: app/views/projects/samples/metadata/file_imports/_dialog.html.erb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                             class:
+                               "w-4 h-4 text-primary-600 bg-slate-100 border-slate-300 rounded focus:ring-primary-500 dark:focus:ring-primary-600 dark:ring-offset-slate-800 focus:ring-2 dark:bg-slate-700 dark:border-slate-600"
+                           },
+                           true,
+                           false %>
+            <%= form.label :ignore_empty_values,
+                       class: "ml-2 text-sm font-medium text-gray-900 dark:text-gray-300" %>
+          </div>
+          <div class="flex flex-row-reverse">
+            <%= form.submit t(".submit_button"),
```
[the remark this answers is not in the document]
Comment (ksierks) [reply app/views/projects/samples/metadata/file_imports/_dialog.html.erb:59]: Should this only be applied when disabled? If so, is you preference to add the classes from the stimulus controller? or apply it for all dropdowns from within the application.css file? Thanks. Please see dbc418133063785e3ce399b53ea0f6f803769c91.

File: app/views/projects/samples/metadata/file_imports/_success.html.erb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<div class="text-center">
+  <%= viral_icon(name: "check_circle_solid", color: :success, classes: "w-16 h-16") %>
```
[the remark this answers is not in the document]
Comment (ksierks) [reply app/views/projects/samples/metadata/file_imports/_success.html.erb]: Made it slightly more transparent in 13f1568befec41241522a5f538f28601272a6d1c.

[the remark this answers is not in the document]
Comment (joshsadam) [reply app/views/projects/samples/metadata/file_imports/_success.html.erb]: I do not like the way that looks, can you please update it to look like the Flowbite version in the link I posted above?

**Review by joshsadam — CHANGES_REQUESTED** [review]
A couple other things.

File: app/views/projects/samples/metadata/file_imports/_dialog.html.erb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            <%= form.file_field :file,
+                            required: true,
+                            accept:
+                              # https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types
+                              "text/csv,application/vnd.ms-excel,application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
+                            data: {
+                              action:
+                                "change->projects--samples--metadata--file-import#readFile"
+                            },
+                            class:
```
[the remark this answers is not in the document]
Comment (ksierks) [reply app/views/projects/samples/metadata/file_imports/_dialog.html.erb:27]: 005affee775b7e10e3851e1b904ec63918d139cc

File: app/views/projects/samples/metadata/file_imports/_success.html.erb (hunk 3 above)
[the remark this answers is not in the document]
Comment (ksierks) [reply app/views/projects/samples/metadata/file_imports/_success.html.erb]: 70bbdd4895b587b544333ec75e65539b37ea33a3 is copied from the above link minus the ok button. I was trying to use the icon component, but I couldn't get an exact match. The closest was wrapping the div around `<%= viral_icon(name: "check", classes: "w-8 h-8 text-green-500 dark:text-green-400") %>`.

File: test/system/projects/samples_test.rb [hunk 5]
```diff
@@ -941,5 +941,52 @@ class SamplesTest < ApplicationSystemTestCase
...
+      within('span[data-controller-connected="true"] dialog') do
+        attach_file 'file_import[file]', Rails.root.join('test/fixtures/files/metadata/mixed_project_samples.csv')
+        find('#file_import_sample_id_column', wait: 1).find(:xpath, 'option[2]').select_option
+        click_on I18n.t('projects.samples.metadata.file_imports.dialog.submit_button')
+      end
+      within %(turbo-frame[id="import_metadata_dialog"]) do
+        assert_text I18n.t('projects.samples.metadata.file_imports.errors.description')
+        click_on I18n.t('projects.samples.metadata.file_imports.errors.ok_button')
+      end
+    end
```
Comment (deepsidhu85) [inline test/system/projects/samples_test.rb:990]: Can you please add tests for these following cases. We have them in the service but should also test them in the UI: 

1. 'duplicate headers'
2. `missing header row`
3. missing metadata column(s)
4. user cannot see the import metadata button if they don't have the correct permissions
5. import metadata with an excel file, 
6. import metadata with an invalid file type
7. import metadata with `ignore empty values` set to `true` and `false`

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
This is looking good. Just have one comment for now to add some testing. Still need to test it out

File: test/system/projects/samples_test.rb [hunk 6]
```diff
@@ -759,7 +759,7 @@ class SamplesTest < ApplicationSystemTestCase
         fill_in I18n.t('projects.samples.attachments.concatenations.modal.basename'), with: 'concatenated file'
         check 'Delete originals'
         click_on I18n.t('projects.samples.attachments.concatenations.modal.submit_button')
-        !assert_html5_inputs_valid # rubocop:disable Lint/Void
+        assert_html5_inputs_valid
```
Comment (deepsidhu85) [inline test/system/projects/samples_test.rb:762]: This line can actually be removed. We check for if the html 5 inputs are valid a few lines below

File: test/system/projects/samples_test.rb (hunk 5 above)
Comment (ksierks) [reply test/system/projects/samples_test.rb:1113]: aad941f0c7f5201d818990e574fa8db57db76bc2

File: test/system/projects/samples_test.rb (hunk 6 above)
Comment (ksierks) [reply test/system/projects/samples_test.rb]: I'm not sure what happened here. If I accidentally deleted something or my IDE auto formatted it. I did not intend to touch this test. 9abbbbd57c12534cbe3fbab885b825516ddf48c0.

**Comment by deepsidhu85** [comment]
So as I was testing this out I realized that the file import service isn't adding any of the errors from the metadata update service to the response. I tried importing metadata to a project with 5 samples that have 2 metadata fields (with values) each, and for one of the samples I updated the metadata provenance for one of the keys to `analysis` using the rails console. Then I tried importing metadata and it shows that the metadata was successfully imported. I see in the file import controller we only check for errors attached to the project. 
To get the error messages attached to the sample by the metadata update service, The `file import service` -> `perform_file_import method` needs to keep track of any errors that were attached to the `sample` from the `metadata update service`. Then before returning the response you can add the `errors` to the response with a key `errors` or something similar. The controller will need to be updated as well to render the appropriate status and send the errors to the UI.

File: test/system/projects/samples_test.rb (hunk 6 above)
Comment (deepsidhu85) [reply test/system/projects/samples_test.rb]: Yeah the IDE probably formatted it!

File: test/system/projects/samples_test.rb [hunk 7]
```diff
@@ -941,5 +940,176 @@ class SamplesTest < ApplicationSystemTestCase
...
+        find('#file_import_sample_id_column', wait: 1).find(:xpath, 'option[2]').select_option
+        click_on I18n.t('projects.samples.metadata.file_imports.dialog.submit_button')
+      end
+      within %(turbo-frame[id="import_metadata_dialog"]) do
+        assert_text I18n.t('projects.samples.metadata.file_imports.success.description')
+        click_on I18n.t('projects.samples.metadata.file_imports.success.ok_button')
+      end
+    end
+
+    test 'should import metadata via invalid file type' do
```
Comment (deepsidhu85) [inline test/system/projects/samples_test.rb:992]: This should read `shouldn't import`

File: test/system/projects/samples_test.rb [hunk 8]
```diff
@@ -941,5 +940,176 @@ class SamplesTest < ApplicationSystemTestCase
...
+      assert_text I18n.t('projects.samples.show.tabs.metadata')
+      click_on I18n.t('projects.samples.show.tabs.metadata')
+      within %(turbo-frame[id="table-listing"]) do
+        assert_text I18n.t('projects.samples.show.table_header.key')
+        assert_selector 'table#metadata-table tbody tr', count: 2
+        assert_no_text 'metadatafield1'
+      end
+    end
+
+    test 'should import metadata with duplicate header errors' do
```
Comment (deepsidhu85) [inline test/system/projects/samples_test.rb:1062]: Same as above

File: test/system/projects/samples_test.rb [hunk 9]
```diff
@@ -941,5 +940,176 @@ class SamplesTest < ApplicationSystemTestCase
...
+        attach_file 'file_import[file]', Rails.root.join('test/fixtures/files/metadata/duplicate_headers.csv')
+        find('#file_import_sample_id_column', wait: 1).find(:xpath, 'option[2]').select_option
+        click_on I18n.t('projects.samples.metadata.file_imports.dialog.submit_button')
+      end
+      within %(turbo-frame[id="import_metadata_dialog_alert"]) do
+        assert_text I18n.t('services.samples.metadata.import_file.duplicate_column_names')
+      end
+    end
+
+    test 'should import metadata with missing metadata row errors' do
```
Comment (deepsidhu85) [inline test/system/projects/samples_test.rb:1075]: Same as above

File: test/system/projects/samples_test.rb [hunk 10]
```diff
@@ -941,5 +940,176 @@ class SamplesTest < ApplicationSystemTestCase
...
+        attach_file 'file_import[file]', Rails.root.join('test/fixtures/files/metadata/missing_metadata_rows.csv')
+        find('#file_import_sample_id_column', wait: 1).find(:xpath, 'option[2]').select_option
+        click_on I18n.t('projects.samples.metadata.file_imports.dialog.submit_button')
+      end
+      within %(turbo-frame[id="import_metadata_dialog_alert"]) do
+        assert_text I18n.t('services.samples.metadata.import_file.missing_metadata_row')
+      end
+    end
+
+    test 'should import metadata with missing metadata column errors' do
```
Comment (deepsidhu85) [inline test/system/projects/samples_test.rb:1088]: Same as above

File: test/system/projects/samples_test.rb [hunk 11]
```diff
@@ -941,5 +940,176 @@ class SamplesTest < ApplicationSystemTestCase
...
+        attach_file 'file_import[file]', Rails.root.join('test/fixtures/files/metadata/missing_metadata_columns.csv')
+        find('#file_import_sample_id_column', wait: 1).find(:xpath, 'option[2]').select_option
+        click_on I18n.t('projects.samples.metadata.file_imports.dialog.submit_button')
+      end
+      within %(turbo-frame[id="import_metadata_dialog_alert"]) do
+        assert_text I18n.t('services.samples.metadata.import_file.missing_metadata_column')
+      end
+    end
+
+    test 'should import metadata with missing sample errors' do
```
Comment (deepsidhu85) [inline test/system/projects/samples_test.rb:1101]: Same as above

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
Few more things for the tests

File: test/system/projects/samples_test.rb (hunk 11 above)
Comment (ksierks) [reply test/system/projects/samples_test.rb]: 6bc47635384870f064716bf2359d79c31156c669

File: test/system/projects/samples_test.rb (hunk 10 above)
Comment (ksierks) [reply test/system/projects/samples_test.rb]: 6bc47635384870f064716bf2359d79c31156c669

File: test/system/projects/samples_test.rb (hunk 9 above)
Comment (ksierks) [reply test/system/projects/samples_test.rb]: 6bc47635384870f064716bf2359d79c31156c669

File: test/system/projects/samples_test.rb (hunk 8 above)
Comment (ksierks) [reply test/system/projects/samples_test.rb]: 6bc47635384870f064716bf2359d79c31156c669

File: test/system/projects/samples_test.rb (hunk 7 above)
Comment (ksierks) [reply test/system/projects/samples_test.rb]: 6bc47635384870f064716bf2359d79c31156c669

**Comment by ksierks** [comment]
> So as I was testing this out I realized that the file import service isn't adding any of the errors from the metadata update service to the response. I tried importing metadata to a project with 5 samples that have 2 metadata fields (with values) each, and for one of the samples I updated the metadata provenance for one of the keys to `analysis` using the rails console. Then I tried importing metadata and it shows that the metadata was successfully imported even though one of the values was not updated as it was set by an analysis. I see in the file import controller we only check for errors attached to the project. To get the error messages attached to the sample by the metadata update service, The `file import service` -> `perform_file_import method` needs to keep track of any errors that were attached to the `sample` from the `metadata update service`. Then before returning the response you can add the `errors` to the response with a key `errors` or something similar. The controller will need to be updated as well to render the appropriate status and send the errors to the UI.

Yeah, I should checking if any metadata is `not_updated` and then append the sample error. From our discussion, I will create a new PR to address this issue as this is fix to the service.

**Review by joshsadam — APPROVED** [review]
This looks good to me, thanks for making that last change to update the table

**Review by ChrisHuynh333 — APPROVED** [review]
This is great and working as described. No comments from me.

**Review by deepsidhu85 — APPROVED** [review]

**Pull request merged**